### PR TITLE
Fix #31: Add docker clean up job

### LINKED ISSUE #31: Add clean up job
We need to clean old containers and images

### PR DESCRIPTION
**Pull request by alexsedova**

### CODE REVIEW

File: dockerizeit/slave/Dockerfile [hunk 1]
```diff
@@ -22,6 +22,9 @@ RUN apk update && apk add py-pip && pip install docker-compose==1.7 docopt==0.6.
 RUN curl --create-dirs -sSLo /usr/share/jenkins/swarm-client-$JENKINS_SWARM_VERSION-jar-with-dependencies.jar http://maven.jenkins-ci.org/content/repositories/releases/org/jenkins-ci/plugins/swarm-client/$JENKINS_SWARM_VERSION/swarm-client-$JENKINS_SWARM_VERSION-jar-with-dependencies.jar \
   && chmod 755 /usr/share/jenkins
 
+#Install tool to do clean up of all unused docker layers and images
+RUN curl -s https://raw.githubusercontent.com/ZZROTDesign/docker-clean/v2.0.4/docker-clean | tee /usr/local/bin/docker-clean > /dev/null && chmod +x /usr/local/bin/docker-clean
```
Comment (Andrey9kin) [inline dockerizeit/slave/Dockerfile]: Would it be possible to import script to the repo? My worry is that we are relying on this gist availability. Also there is seems to be no version so we will not spot if it changes

Comment (Andrey9kin) [inline dockerizeit/slave/Dockerfile]: Another option is to look up on GitHub - perhaps it exists as repo? Then we could fork it to Praqma and use it from there

Comment (alexsedova) [inline dockerizeit/slave/Dockerfile]: There is the repo:  https://github.com/ZZROTDesign/docker-clean I'm sure we are allowed to import the script - it won't be fair. As well as fork it and use it by Praqma. We could fork it and keep just in a case if something goes wrong with original script. Another option - fork and contribute.

Comment (Andrey9kin) [inline dockerizeit/slave/Dockerfile]: @alexsedova let's fork it to make sure that we have a copy and then use that copy

File: jobdsl-gradle/src/main/groovy/job/JobBuilder.groovy [hunk 2]
```diff
@@ -284,6 +301,31 @@ public class JobBuilder {
...
+     * and https://wiki.jenkins-ci.org/display/JENKINS/NodeLabel+Parameter+Plugin
+     * There is a reference to issue [JENKINS-33756](https://issues.jenkins-ci.org/browse/JENKINS-33756)
+     * "Run on all nodes matching the label" or allNodes in DSL flag schedules one more build
+     * on the node job has been started.
+     *
+     * @String paramName                  the paramater name
+     * @ArrayList<String> nodesLabels     list of labels of nodes you want to run the job
+     *
+     */
+    public JobBuilder addJobBuildParameter(String paramName, ArrayList<String> nodesLabels, boolean debug = false) {
```
Comment (Andrey9kin) [inline jobdsl-gradle/src/main/groovy/job/JobBuilder.groovy]: This is a too generic name - you are adding a very specific type of parameter. BTW, do we really need to wrap such one-liners? Would it be possible to just update the object when it is created

File: jobdsl-gradle/src/main/groovy/job/JobBuilder.groovy [hunk 3]
```diff
@@ -172,7 +172,24 @@ public class JobBuilder {
...
     }
-
+    /**
+     * Use this function to add Build trigger
+     *
+     * @ArrayList<String>    projects List of upsteam projects trigger will watch
+     * @String               threshold the result of the upstreams build when run the job.
+     *                       SUCCESS by default
+     */
+        public JobBuilder addBuildTrigger(ArrayList<String> projects, String threshold = 'SUCCESS', boolean debug = false) {
```
Comment (Andrey9kin) [inline jobdsl-gradle/src/main/groovy/job/JobBuilder.groovy]: This is a too generic name - you are adding a very specific type of trigger. BTW, do we really need to wrap such one-liners? Would it be possible to just update the object when it is created

File: jobdsl-gradle/src/jobs/groovy/dockercleanup.groovy [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ * If we have 4 slaves are matching one label, we will have 5 running jobs instead of 4.
+ * These issue is not resolved yet.
+ * We don't use cluster of docker hosts, that is why we are running job on
+ * all slaves matching "docker" label even they are running on the one host. We could run
+ * clean up job just once, while we don't have a multihost cluster - it is another
+ * possible solution. In this case we will not get the issue with double job running.
+ **/
+new JobBuilder(this as DslFactory, "jenkins_as_a_code-cleaup-docker")
+    .addLogRotator()
+    .addBuildTrigger(["jenkins_as_a_code-pipeline"])
```
Comment (Andrey9kin) [inline jobdsl-gradle/src/jobs/groovy/dockercleanup.groovy]: I have a concern - we are about to trigger clean-up job right after the pipeline. At the same time Munchausen is about to kill Jenkins master in 15 sec after pipeline ends. Perhaps we could use a time trigger and then call it 3 am or something like that instead

**Comment by Andrey9kin** [comment]
@alexsedova was just about to merge this one but few comments crossed my mind. Sorry for not spotting that earlier

File: jobdsl-gradle/src/main/groovy/job/JobBuilder.groovy (hunk 3 above)
Comment (alexsedova) [inline jobdsl-gradle/src/main/groovy/job/JobBuilder.groovy]: I could add switch-case to cover more cases. Add todo.

**Pull request merged**